xsuite/src/cli/buildReproducibleAction.ts

```diff
@@ -0,0 +1,205 @@
+import { PathLike, existsSync, mkdirSync, readdirSync } from "fs";
+import path from "path";
+import { Command } from "commander";
+import { log } from "../_stdio";
+import {
+  defaultIgnore,
+  findBuildableContractDirs,
+  getGid,
+  getUid,
+} from "./buildUtils";
+import {
+  defaultReproducibleDockerImage,
+  logAndRunCommand,
+  logError,
+  promptUserWithRetry,
+} from "./helpers";
+
+export const registerBuildReproducibleCmd = (cmd: Command) => {
+  cmd
+    .command("build-reproducible")
+    .description("Build contract.")
+    .option(
+      "--image <IMAGE_TAG>",
+      "Specify the tag of the docker image, that is used to build the contract (e. g. multiversx/sdk-rust-contract-builder:v6.1.0)",
+    )
+    .option(
+      "--dir <DIR>",
+      "Directory in which the command is executed (default: $(PWD))",
+    )
+    .option(
+      "--target-dir <TARGET_DIR>",
+      "Target directory used by Rust compiler (default: $(PWD)/target)",
+    )
+    .option(
+      "--ignore <IGNORE>",
+      `Ignore all directories matching the RegExp (default: ${defaultIgnore})`,
+    )
+    .option("-r, --recursive", "Build all contracts under the directory")
```

Comment:
When I run `pnpm xsuite build-reproducible -r --target-dir $(pwd)/output-reproducible` inside of the folder `xsuite/contracts`, I get the following output files:

![image](https://github.com/arda-org/xSuite/assets/5437552/6ac20963-b68b-4aa8-a9d6-32b707ec68e2)

There are 2 different `artifacts.json` files, 2 `data` folders and 2 `world` folders. This file structure is unnecessarily complex and we just aim for simply this:

```
output-reproducible/
  data/
    data.wasm
    ...
  world/
    data.wasm
    ...
  artifacts.json
```

We get this complex file structure because, to my understanding, the MultiversX Docker image builds recursively by default and assume multiple contracts will be present by default.

So I think we should try to get the following behavior:

If the flag `-r` or `--recursive` is provided, we call only once the MvX Docker image (we don't call it once per contract). Regarding the output files structure, @janniksam do you think we should keep the one outputted by the MvX Docker image or we should rather try to have one `output-reproducible` per smart contract folder?

If the flag `-r` or `--recursive` is not provided, we call once the MvX Docker image without the recursive mode (I need to double check with the team if this is possible). Regarding the output files structure, it would be best if it doesn't have multiple level of folders, e.g. we don't want:

```
output-reproducible/
  contract/
    contract.wasm
  artifacts.json
```

but rather:

```
output-reproducible/
  artifacts.json
  contract.wasm
```

Finally, it would be ideal to have tests that verify that the output files structure is correct (one test for the recursive case on `xsuite/contracts`, one test for the non-recursive case on `contracts/blank`).

Reply:
should be better now :)